client/client.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+
+	"github.com/GeorgeLuo/grpc/models"
+)
+
+// TODO persistent connection
+
+var client *http.Client
+
+func main() {
+
+	if len(os.Args) < 2 {
+		fmt.Println("start, stop, or command instruction not provided")
+		os.Exit(1)
+	}
+
+	args := os.Args
+	var certFile string
+	var keyFile string
+	var host string
+	var endpoint string
+
+	body, err := ReadArgs(args, &endpoint, &certFile, &keyFile, &host)
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	var request *http.Request
+
+	switch endpoint {
+	case "start":
+		request, err = StartRequest(body, host)
+	case "stop":
+		request, err = StopRequest(body, host)
+	case "status":
+		request, err = StatusRequest(body, host)
+	default:
+		err = errors.New("invalid command")
+		os.Exit(1)
+	}
+
+	client, err = newClient(certFile, keyFile, certFile)
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	r, err := client.Do(request)
+	if err != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	defer r.Body.Close()
+	responseBody, responseError := ioutil.ReadAll(r.Body)
+	if responseError != nil {
+		fmt.Println(err.Error())
+		os.Exit(1)
+	}
+
+	fmt.Printf("%s", responseBody)
+}
+
+// newClient creates a new tls client from key and certs
+func newClient(certFile string, keyFile string, caCertFile string) (*http.Client, error) {
+
+	cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO add command line arg for cacert
+	caCert, err := ioutil.ReadFile(caCertFile)
+	if err != nil {
+		return nil, err
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	return &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: &tls.Config{
+				RootCAs:      caCertPool,
+				Certificates: []tls.Certificate{cert},
+			},
+		},
+	}, nil
+}
+
+// TODO could return a config object holding this information
+
+// ReadArgs populates the common parameters between the 3 endpoints and returns request interface.
+func ReadArgs(args []string, endpoint *string, certFile *string, keyFile *string, host *string) (body interface{}, err error) {
```

Comment:
I see. I wished not to repeat the cert and key flags for each of the subcommands so that's why I didn't define all the flagsets. This probably was reductionism at the expense of flexibility.